rewrite functions using new api

applewiskey opened the pull request:
Rewrite functions using new function API
    Fixes #321 
    Fixes #320 
    Fixes #225 (?)

bastelfreak commented:
Hi @applewiskey, thanks for the PR. Can you take a look at the failing tests on travis?

alexjfisher commented on lib/puppet/functions/archive/go_md5.rb:
```diff
@@ -0,0 +1,32 @@
+# args[0] - username
+# args[1] - password
+# args[2] - file_name
+# args[3] - go md5 checksum url
```
Could you convert this documentation to the new puppet strings format?  eg like https://github.com/voxpupuli/puppet-archive/blob/ab996a12193497c13c3a2e151915652289cd1e0b/lib/puppet/functions/archive/artifactory_checksum.rb#L5-L9

alexjfisher commented:
@applewiskey Thanks for taking this on!  The tests will need moving/updating.  This shouldn't be too tricky.  Take a look at https://github.com/voxpupuli/puppet-archive/pull/323/files and ping me if you get stuck and need help.

Will the manifests that call these functions will also need updating to use the new namespace?

ekohl commented on lib/puppet/functions/archive/go_md5.rb:
```diff
@@ -0,0 +1,32 @@
+    required_param 'String', :username
+    required_param 'String', :password
+    required_param 'String', :file
+    required_param 'String', :url
```
Should this use `Stdlib::Httpurl` or are protocols other than http(s) also supported?

applewiskey replied:
It checked against String in define archive::nexus anyway

ekohl commented on lib/puppet/functions/archive/assemble_nexus_url.rb:
```diff
@@ -0,0 +1,15 @@
+  end
+
+  def assemble_nexus_url(nexus_url, params)
+    service_relative_url = 'service/local/artifact/maven/content'
```
Should this be a param with a default too?

applewiskey replied:
I don't it should.

applewiskey commented:
Sorry for a delay.
I don't get why first 3 tests failing and what should I come up with reading this
```
      Failure/Error: Puppet.features.stubs(:root?).returns(true)
      
      Mocha::NotInitializedError:
        Mocha methods cannot be used outside the context of a test
```

Anyway, it doesn't fix #320

bastelfreak commented:
@applewiskey the issues should go away after a rebase. They are caused by a specific version of the mocha gem.

ekohl commented:
Could you rebase and fix the rubocop error?

bastelfreak commented:
I get the following errors locally after rebasing:

```
Failures:

  1) go_md5 retreives file md5
     Failure/Error: expect(scope.function_go_md5(['user', 'pass', 'filea', url])).to eq '283158c7da8c0ada74502794fa8745eb'
     
     NoMethodError:
       undefined method `function_go_md5' for Scope():Puppet::Parser::Scope
     # ./.vendor/ruby/2.5.0/gems/puppet-5.5.3/lib/puppet/parser/scope.rb:1037:in `method_missing'
     # ./spec/unit/puppet/parser/functions/go_md5_spec.rb:16:in `block (2 levels) in <top (required)>'
```

ekohl commented:
Also adding the backwards-incompatible flag because this removes the old functions and users could rely on that. I'm leaning to keeping them in for now so it's not a major version bump.

bastelfreak commented:
ping @applewiskey :)

applewiskey commented:
Rebased. Anything else?

bastelfreak commented:
@applewiskey please take a look at the failing travis jobs.

applewiskey commented:
I fixed rubocop, but I don't really know how to fix something like this
```
F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 1)
F    should contain File[/opt/app/example.zip] with owner => "0", group => "0", mode => "0640" and require => "Archive[/opt/app/example.zip]" (FAILED - 2)
  go archive with path
F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 3)
F    should contain File[/opt/app/example.zip] with owner => "app", group => "app", mode => "0400" and require => "Archive[/opt/app/example.zip]" (FAILED - 4)
```
and a lot more different errors. Errors not in code I changed.

alexjfisher commented:
> I fixed rubocop, but I don't really know how to fix something like this
> 
> ```
> F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 1)
> F    should contain File[/opt/app/example.zip] with owner => "0", group => "0", mode => "0640" and require => "Archive[/opt/app/example.zip]" (FAILED - 2)
>   go archive with path
> F    should contain Archive[/opt/app/example.zip] with path => "/opt/app/example.zip", source => "http://home.lan:8081/go/example.zip", checksum => "0d4f4b4b039c10917cfc49f6f6be71e4" and checksum_type => "md5" (FAILED - 3)
> F    should contain File[/opt/app/example.zip] with owner => "app", group => "app", mode => "0400" and require => "Archive[/opt/app/example.zip]" (FAILED - 4)
> ```
> 
> and a lot more different errors. Errors not in code I changed.

The spec test for the defined type was mocking the old non-namespaced function.  This needed to be [updated](https://github.com/voxpupuli/puppet-archive/pull/392/files#diff-a6dbe892af312a9f108f66d707514aa9L6-L10).

@applewiskey Thanks for your work on this, and sorry we didn't follow it through and get it merged at the time.

I've superceded this PR with 3 smaller ones.  I think I've covered everything you had, as well as getting the tests converted and passing.
https://github.com/voxpupuli/puppet-archive/pull/390
https://github.com/voxpupuli/puppet-archive/pull/391
https://github.com/voxpupuli/puppet-archive/pull/392

applewiskey commented:
Closing as there are
#390
#391
#392
by @alexjfisher
